crypto/core_namemap.c

```diff
@@ -91,14 +97,21 @@ static const OPENSSL_CTX_METHOD stored_namemap_method = {
 
 int ossl_namemap_empty(OSSL_NAMEMAP *namemap)
 {
-    int rv = 0;
+#ifdef tsan_ld_acq
+    /* Have TSAN support */
+    return namemap == NULL || tsan_load(&namemap->max_number) == 0;
+#else
+    /* No TSAN support */
```

Comment:
So this means we still have to be careful about possible locking issues, but that we might not catch that issue because most of us (as well as Travis) run our tests in an environment where we have access to atomics and thereby have "proper" TSAN support...  While I like this change, that's something that we should be *very* clear about.

Reply:
Seriously?????????????  You were the person who pointed out this exact problem in my previous PR.  We've used TSAN primitives without locking for a while now.  They aren't safe.

As a solution, I think we should wrapper up the TSAN code (in a struct) so that it locks if the TSAN calls are not available.  Currently, we use TSAN for _int_, _unsigned int_ and _unsigned long int_.  The final flavour will disappear with @richsalz's hash changes.  Still, this needs to be sorted out before 3.0.

Reply:
I haven't been aware of TSAN before I saw your changes in that other PR.  Looking at the TSAN macros now, I find it striking how much (not exactly, but very much) they do the same thing as the CRYPTO_UP_REF / CRYPTO_DOWN_REF functions.  As a matter of fact, we could easily reimplement much of the latter in TSAN terms.

The fact remains, though, that the more we let locking mechanisms be for "legacy" platforms only, the more we risk getting bug reports that we have a hard time debugging.  We might want to consider a knob that turns off our use of atomics for that very purpose.

Reply:
The legacy locking is the reason I'd like to wrap these up more nicely.  A user (i.e. us) shouldn't care if locks need to be used or not.